src/safeds_stubgen/api_analyzer/_ast_visitor.py

```diff
@@ -263,10 +273,78 @@ def enter_funcdef(self, node: mp_nodes.FuncDef) -> None:
                 # Sort for the snapshot tests
                 type_var_types.sort(key=lambda x: x.name)
 
+        # Check docstring parameter types vs code parameter type hint
+        for i, argument in enumerate(arguments):
+            arg_type = argument.type
+            doc_type = argument.docstring.type
+
+            if arg_type is None and doc_type is not None:
+                # No code type hint but docstring type
+                arguments[i] = Parameter(
+                    id=argument.id,
+                    name=argument.name,
+                    is_optional=argument.docstring.default_value != "",
+                    default_value=argument.docstring.default_value,
```

Comment:
The default value should always be taken from the code. Using the docstring would require additional validation.

Likewise, we can always determine from the code whether a parameter is optional.

Reply:
Instead of creating a new parameter object, it should suffice to replace the type using [`dataclasses.replace`](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace).

Then we can also combine some branches:

```py
            if (
                code_type is not None
                and doc_type is not None
                and code_type != doc_type
                and self.type_source_warning == TypeSourceWarning.WARN
            ):
                msg = f"Different type hint and docstring types for '{function_id}'."
                logging.warning(msg)

            if doc_type is not None and (
                code_type is None
                or self.type_source_preference == TypeSourcePreference.DOCSTRING
            ):
                parameters[i] = dataclasses.replace(
                    parameter,
                    type=doc_type,
                )
```